shopinvader/services/abstract_download.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import mimetypes
+
+from odoo import _
+from odoo.addons.base_rest.components.service import (
+    skip_secure_response,
+    to_int,
+)
+from odoo.addons.component.core import AbstractComponent
+from odoo.exceptions import MissingError
+from odoo.http import content_disposition, request
+
+
+class AbstractDownload(AbstractComponent):
+    """
+    Class used to define behaviour to generate and download a document.
+    You only have to inherit this AbstractComponent and implement the function
+    _get_report_action(...).
+
+    Example for invoice service:
+    Class InvoiceService(Component):
+        _inherit = [
+            "base.shopinvader.service",
+            "abstract.shopinvader.download",
+        ]
+        _name = "shopinvader.invoice.service"
+
+        def _get_report_action(self, target):
+            return target.invoice_print()
+    """
+
+    _name = "abstract.shopinvader.download"
+
+    # This function should be overwritten
+    def _get_report_action(self, target):
```

Comment:
I'd pass `params` here to be able to control the action on demand.
Not a blocker tho,